homeassistant/components/screenlogic/__init__.py

```diff
@@ -140,38 +135,34 @@ def __init__(self, hass, *, config_entry, gateway, api_lock):
             _LOGGER,
             name=DOMAIN,
             update_interval=interval,
-            # We don't want an immediate refresh since the device
-            # takes a moment to reflect the state change
+            # Debounced option since the device takes
+            # a moment to reflect the knock-on changes
             request_refresh_debouncer=Debouncer(
                 hass, _LOGGER, cooldown=REQUEST_REFRESH_DELAY, immediate=False
             ),
         )
 
-    def reconnect_gateway(self):
-        """Instantiate a new ScreenLogicGateway, connect to it and update. Return new gateway to caller."""
-
-        connect_info = get_connect_info(self.hass, self.config_entry)
-
-        try:
-            gateway = ScreenLogicGateway(**connect_info)
-            gateway.update()
-        except ScreenLogicError as error:
-            raise UpdateFailed(error) from error
-
-        return gateway
-
     async def _async_update_data(self):
         """Fetch data from the Screenlogic gateway."""
         try:
-            async with self.api_lock:
-                await self.hass.async_add_executor_job(self.gateway.update)
+            await self.gateway.async_update()
         except ScreenLogicError as error:
-            _LOGGER.warning("ScreenLogicError - attempting reconnect: %s", error)
+            _LOGGER.warning("Update error - attempting reconnect: %s", error)
+
+            # Clean up the previous connection as we're about to create a new one
+            await self.gateway.async_disconnect()
 
-            async with self.api_lock:
-                self.gateway = await self.hass.async_add_executor_job(
-                    self.reconnect_gateway
-                )
+            connect_info = await async_get_connect_info(self.hass, self.config_entry)
+
+            self.gateway = ScreenLogicGateway(**connect_info)
+
+            try:
+                await self.gateway.async_update()
+            except (ScreenLogicError, ScreenLogicWarning) as ex:
+                raise UpdateFailed(ex) from ex
```

Comment:
Would be a bit cleaner if this was a new function .. maybe `_async_reconnect_update_data`

Reply:
https://github.com/home-assistant/core/pull/60466/commits/f2a27c7ea7ae5cd43b9ccd488588a036d35b3266